server/src-lib/Hasura/Server/Auth/Utils.hs

```diff
@@ -0,0 +1,52 @@
+module Hasura.Server.Auth.Utils
+  ( timeFromCacheControlHeader
+  , timeFromExpiresHeader
+  , toUTCTime
+  , toNominalDiffTime
+  ) where
+
+import           Data.Parser.CacheControl (parseMaxAge)
+import           Data.Time.Clock
+import           Data.Time.Format         (defaultTimeLocale, parseTimeM)
+
+import qualified Data.String.Conversions  as CS
+import qualified Data.Text                as T
+
+import           Hasura.Prelude
+
+
+data ExpiryTime = AbsoluteTime UTCTime
+                | RelativeTime NominalDiffTime
+
+
+-- | Extracts a relative expiry time from the CacheControl header.
+timeFromCacheControlHeader
+  :: (Monad m)
+  => Maybe T.Text         -- ^ the header
+  -> (T.Text -> m ())     -- ^ what should be done if parsing fails
+  -> m (Maybe ExpiryTime)
+timeFromCacheControlHeader header onError =
+  onMaybe header $ \h -> case parseMaxAge h of
+    Left e       -> onError (T.pack e) >> return Nothing
+    Right maxAge -> return $ Just $ RelativeTime $ fromInteger maxAge
```

Comment:
This converts the result of `parseMaxAge` to a `NominalDiffTime`, which is good! But arguably `parseMaxAge` should just return a `NominalDiffTime` in the first place, so I think you can eliminate this function completely and change the type of `parseMaxAge` directly.

Reply:
I didn't touch parseMaxAge in the end, but simply `fmap fromInteger` at the call site, which I think is acceptable?